airbyte-integrations/integration-test-lib/src/main/java/io/airbyte/integrations/base/TestDestination.java

```diff
@@ -0,0 +1,243 @@
+    // todo (cgardens) - blocked on worker not calling discover worker.
+  }
+
+  /**
+   * Verify that when given invalid credentials, that check connection returns a failed response.
+   * Assume that the {@link TestDestination#getInvalidConfig()} is invalid.
+   */
+  @Test
+  void testCheckConnectionInvalidCredentials() {
+    // todo (cgardens) - blocked on worker not calling discover worker.
+  }
+
+  private static class DataArgumentsProvider implements ArgumentsProvider {
+
+    @Override
+    public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
+      return Stream.of(
+          Arguments.of("exchange_rate_messages.txt", "exchange_rate_catalog.json")
+      // todo - need to use the new protocol to capture this.
+      // Arguments.of("stripe_messages.txt", "stripe_schema.json")
+      );
+    }
+
+  }
+
+  /**
+   * Verify that the integration successfully writes records. Tests a wide variety of messages and
+   * schemas (aspirationally, anyway).
+   */
+  @ParameterizedTest
+  @ArgumentsSource(DataArgumentsProvider.class)
+  void testSync(String messagesFilename, String catalogFilename) throws Exception {
+    final Schema catalog = Jsons.deserialize(MoreResources.readResource(catalogFilename), Schema.class);
+    final List<SingerMessage> messages = MoreResources.readResource(messagesFilename).lines()
+        .map(record -> Jsons.deserialize(record, SingerMessage.class)).collect(Collectors.toList());
+    runSync(messages, catalog);
+
+    assertSameMessages(messages, recordRetriever(testEnv));
+  }
+
+  /**
+   * Verify that the integration overwrites the first sync with the second sync.
+   */
+  @Test
+  void testSecondSync() throws Exception {
+    final Schema catalog = Jsons.deserialize(MoreResources.readResource("exchange_rate_catalog.json"), Schema.class);
+    final List<SingerMessage> firstSyncMessages = MoreResources.readResource("exchange_rate_messages.txt").lines()
+        .map(record -> Jsons.deserialize(record, SingerMessage.class)).collect(Collectors.toList());
+    runSync(firstSyncMessages, catalog);
+
+    List<SingerMessage> secondSyncMessages = Lists.newArrayList(new SingerMessage()
+        .withStream(catalog.getStreams().get(0).getName())
+        .withRecord(OBJECT_MAPPER.createObjectNode()
+            .put("date", "2020-03-31T00:00:00Z")
+            .put("HKD", 10)
+            .put("NZD", 700)));
+    runSync(secondSyncMessages, catalog);
+    assertSameMessages(secondSyncMessages, recordRetriever(testEnv));
+  }
+
+  private void runSync(List<SingerMessage> messages, Schema catalog) throws IOException, WorkerException {
+    final StandardTargetConfig targetConfig = new StandardTargetConfig()
+        .withDestinationConnectionImplementation(new DestinationConnectionImplementation().withConfiguration(getConfig()))
+        .withStandardSync(new StandardSync().withSchema(catalog));
+
+    final DefaultSingerTarget target = new DefaultSingerTarget(getImageName(), pbf);
```

Comment:
now that integrations are fully self contained, why not test against the docker image directly? that way, integration packages are more isolated from the core codebase internals which gives us more freedom to make change in core

Reply:
last time we went to launch, we did what you described and had test on the output of the dockerfiles. then in the 11th hour we found that most of the integrations didn't really work with the workers. this was because the tests on the output of the dockerfiles didn't actually test that what was output was compatible with the workers. 

now i think you could make the counterargument that that was because the tests were just not written against the proper types. but i got burned and had to scramble alot, so i'm trying to give highest guarantee that stuff actually works here and am pretty willing to tradeoff some isolation for it.